File: QytetetRuby/lib/especulador.rb
#Elena Cantero Molina
#encoding: utf-8
#trig.rb

module ModeloQytetet
  class Especulador < Jugador
    attr_reader :fianza
    
    def initialize(jugador, fianza)
      super(jugador.nombre)
      @fianza = fianza
      @factor_especulador = 2
    end
    
    def factor_especulador
      return @factor_especulador
    end
    
    def pagar_impuestos(cantidad)
      modificar_saldo(-cantidad/2)
    end
    
    def convertirme(fianza)
      return self
    end
    
    def ir_a_carcel(casilla)
      if !pagar_fianza(@fianza)
        @casilla_actual = casilla
        @encarcelado = true
      end
    end
    
    def pagar_fianza(cantidad)
      if @saldo > @fianza
        modificar_saldo(-@fianza)
        return true
      end
      return false
    end
    
    def to_s
      super
    end
    
    protected :pagar_impuestos, :ir_a_carcel, :convertirme
    private :pagar_fianza
  end
end

File: QytetetJava/src/modeloqytetet/Especulador.java
/**
 * @author Elena Cantero Molina
 */
package modeloqytetet;

public class Especulador extends Jugador{
    static int factorEspeculador = 2;
    int fianza;
        
    protected Especulador(Jugador jugador, int fianza){
        super(jugador);
        this.fianza = fianza;
    }
    
    @Override
    protected void pagarImpuestos(int cantidad){
        modificarSaldo(-(cantidad/2));
    }
    
    @Override
    protected void irACarcel(Casilla casilla){
        if(pagarFianza(fianza)){
            modificarSaldo(-fianza);
        }
        else{
            setCasillaActual(casilla);
            setEncarcelado(true);
        }
    }
    
    @Override
    protected Especulador convertirme(int fianza){ return this; }
    
    boolean pagarFianza(int cantidad){return cantidad < getSaldo(); }

    @Override
    public String toString() {
        return "Especulador{" +super.toString()+ "fianza=" + fianza + "}\n";
    }
}
